finality-aleph/src/validator_network/manager.rs

```diff
@@ -0,0 +1,237 @@
+    incoming: HashMap<AuthorityId, oneshot::Sender<()>>,
+}
+
+/// Error during sending data through the Manager
+#[derive(Debug, PartialEq, Eq)]
+pub enum SendError {
+    /// Outgoing network connection closed
+    ConnectionClosed,
+    /// Peer not added to the manager
+    PeerNotFound,
+}
+
+impl Display for SendError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use SendError::*;
+        match self {
+            ConnectionClosed => write!(f, "worker dead"),
+            PeerNotFound => write!(f, "peer not found"),
+        }
+    }
+}
+
+struct ManagerStatus {
+    wanted_peers: usize,
+    incoming_peers: usize,
+    outgoing_peers: usize,
+}
+
+impl Display for ManagerStatus {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        if self.wanted_peers == 0 {
+            return write!(f, "not maintaining any connections");
+        }
+        if self.incoming_peers == 0 {
+            write!(f, "WARNING! No incoming peers even though we expected tham, maybe connecting to us is impossible.")?;
+        }
+        write!(
+            f,
+            "maintaining {} connections, incoming connections {}, outgoing connections {}",
+            self.wanted_peers, self.incoming_peers, self.outgoing_peers,
+        )
+    }
+}
+
+impl<A: Data, D: Data> Manager<A, D> {
+    /// Create a new Manager with empty list of peers.
+    pub fn new() -> Self {
+        Manager {
+            addresses: HashMap::new(),
+            outgoing: HashMap::new(),
+            incoming: HashMap::new(),
+        }
+    }
+
+    /// Add a peer to the list of peers we want to stay connected to, or
+    /// update the list of addresses if the peer was already added.
+    /// Returns whether this peer is a new peer.
+    pub fn add_peer(&mut self, peer_id: AuthorityId, addresses: Vec<A>) -> bool {
+        self.addresses.insert(peer_id, addresses).is_none()
+    }
+
+    /// Return Option containing addresses of the given peer, or None if
+    /// the peer is unknown.
+    pub fn peer_addresses(&self, peer_id: &AuthorityId) -> Option<Vec<A>> {
+        self.addresses.get(peer_id).cloned()
+    }
+
+    /// Add an established outgoing connection with a known peer,
+    /// but only if the peer is on the list of peers that we want to stay connected with.
+    pub fn add_outgoing(
+        &mut self,
+        peer_id: AuthorityId,
+        data_for_network: mpsc::UnboundedSender<D>,
+    ) {
+        if self.addresses.contains_key(&peer_id) {
+            self.outgoing.insert(peer_id, data_for_network);
+        };
+    }
+
+    /// Add an established incoming connection with a known peer,
+    /// but only if the peer is on the list of peers that we want to stay connected with.
+    pub fn add_incoming(&mut self, peer_id: AuthorityId, exit: oneshot::Sender<()>) {
+        if self.addresses.contains_key(&peer_id) {
+            self.incoming.insert(peer_id, exit);
```

Comment:
I think it would be good to log something in case we overwrite an incoming connection -- it is something unexpected that we can miss.

Reply:
Added.